support/cas-server-support-openrewrite/build.gradle

```diff
@@ -1,38 +1,37 @@
 description = "Apereo CAS Open Rewrite Support"
 
-processResources {
-    def casVersion = project.version
+
+tasks.register("createRecipe") {
+    def casVersion = rootProject.version
     def tomcatVersion = project.tomcatVersion
     def sourceCompatibility = project.sourceCompatibility
     def springBootVersion = project.springBootVersion
     def jibVersion = project.jibVersion
     def gradleVersion = gradle.gradleVersion
     def recipeName = "cas${casVersion.replaceAll(/\.|-SNAPSHOT/, '')}"
     def templateContents = file("src/main/resources/current.gtemplate").text
-
-    def outputDir = file("src/main/resources/META-INF/rewrite")
-    def outputFile = file("${outputDir.absolutePath}/${recipeName}.yml")
-    outputs.file(outputFile)
+    def recipeFileName = "$buildDir/generated/META-INF/rewrite/${recipeName}.yml"
 
     doFirst {
         def model = [:]
-
         model.put("recipeName", recipeName)
         model.put("tomcatVersion", tomcatVersion)
         model.put("casVersion", casVersion)
         model.put("javaVersion", sourceCompatibility)
         model.put("jibVersion", jibVersion)
         model.put("gradleVersion", gradleVersion)
         model.put("springBootVersion", springBootVersion)
-        
+
         def engine = new groovy.text.GStringTemplateEngine()
         def recipe = engine.createTemplate(templateContents).make(model).toString()
 
-        if (!outputDir.exists()) {
-            outputDir.mkdirs()
-        }
+        def outputFile = file(recipeFileName)
+        outputFile.parentFile.mkdirs()
```

Comment:
Thanks for the suggestions!

If what I have here (latest commit) is in fact correct, I see:

```
* What went wrong:
Execution failed for task ':support:cas-server-support-openrewrite:createRecipe'.
> java.io.FileNotFoundException: /Users/misagh/cas-server/support/cas-server-support-openrewrite/build/generated/META-INF/rewrite/cas700.yml (No such file or directory)
```

Are the most recent changes what you had in mind?

Reply:
@mmoayyed in the current version, you are missing the line `outputs.file(outputFile)` which wires that file to the tasks's output which is why you are seeing the `FileNotFoundException`

Reply:
This is not related to the configuration cache. It is a build caching issue. You will also see this same warning in the deprecations section of your build scan.

I didn't check to see if this is the case, but it's possible that multiple tasks are writing to the `generated` directory. Each task requires an exclusive output directory in order to guarantee correctness. Does the problem go away if you write to a different directory? For example `createRecipe-generated`?

Reply:
Let's say you're presently running CAS 7.0.0, have fallen behind a bit and now wish to upgrade to 7.0.5. The recipe that allows you to upgrade from 7.0.0 should ideally allow you to go to 7.0.1 first, and then 7.0.2, and so on until you get to 7.0.5. Essentially the JAR that contains upgrade instructions in form of recipes should allow you to travel between versions separately, and as a result, needs to track the version history so you can choose where to go. 

There is of course no hard requirement to make that many stops. In *almost* all scenarios, one can directly go to 7.0.5, and other than the software version itself, (specially in patch release), all such recipes are completely identical. So this might be somewhat overkill and strictly speaking, history as such does not need to be kept around. But on the off chance that the recipes begin to diverge and the steps for going from A to B start to involve more than just a simple version change (i.e. files need to move around, other libs updates, properties renamed, etc), it seems ideal to isolate and track recipes in history and let folks choose where to go.

Another use case for keeping history is waiting for or backing out for bug fixes. If 7.0.5 has a major defect, you likely want to step back to 7.0.4 or 7.0.3 and wait until 7.0.6 or later solves that issue. To do this, recipe files and history need to be available in history to let you navigate.

In the world of OpenRewrite (OR), one gets to produce a JAR that contains such recipes and these get discovered by the OR Gradle plugin. So if you are on version 7.0.0, in theory and practice the build offers you options to download the OR JAR for version 7.0.X which allows you to navigate and upgrade one step at a time until you get to 7.0.X. This means that just before 7.0.X is released, we'd have to run this custom task, create and curate the new recipe for the upgrade, commit it, and release which would then allow the adopter/user to download and upgrade with history in place. The OR JAR would contain that newly made recipe plus everything else before it. You as the adopter would only do this download once. You get the OR JAR for 7.0.X and you can start moving around or go back as far as the big bang so long as recipes are available! If we do not keep previous recipes, that means every OR JAR would only contain one recipe for the version it's tied to, and that means multiple downloads/attempts by the user for every version. 

PS One could perfectly argue that such recipes in this OR JAR should be moved to a separate project to not be tied to the version of the software that is being developed here in this branch. While perfectly true, that would seem to produce a lot more maintenance overhead.

Reply:
Actually, after a discussion with @runningcode , a better solution would be to put the generated rewrite yaml files in a directory outside of build or source dirs - that way there's no clashing of task outputs writing to a directory that's also that tasks input.